New servus extractor logic #1076

FrankZ85 opened the pull request:
<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### Description of your *pull request* and other information

</details>

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

DESCRIPTION

Fixes #1076

Ashish0804 commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,114 +12,58 @@ class ServusIE(InfoExtractor):
-                })
+            'https://api-player.redbull.com/stv/servus-tv?timeZone=Europe/Berlin&videoId=%s' % video_id,
+            video_id, 'Downloading video JSON')
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
+            video.get('videoUrl'), video_id, 'mp4', entry_protocol='m3u8_native',
+            m3u8_id='hls', fatal=False)
```
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            video['videoUrl'], video_id, 'mp4', entry_protocol='m3u8_native',
            m3u8_id='hls')
```

gamer191 commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,114 +12,58 @@ class ServusIE(InfoExtractor):
         'only_matching': True,
     }]
 
+
```
The blank line at line 48 should be removed, in order to make the automated linter happy
```suggestion

```

StefanLobbenmeier commented:
I just ran a successful test of this on my machine. The old version fails like this:
```
➜  Downloads ./yt-dlp_macos_legacy https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
[generic] aa7u8970e8drj41nx6z0: Requesting header
WARNING: [generic] Falling back on generic information extractor.
[generic] aa7u8970e8drj41nx6z0: Downloading webpage
[generic] aa7u8970e8drj41nx6z0: Extracting information
ERROR: Unsupported URL: https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
```

After applying this patch I could successfully download:
➜  Downloads ./yt-dlp_macos_legacy-2 https://www.servustv.com/sport/v/aa7u8970e8drj41nx6z0/
```
[Servus] AA7U8970E8DRJ41NX6Z0: Downloading video JSON
[Servus] AA7U8970E8DRJ41NX6Z0: Downloading m3u8 information
[info] AA7U8970E8DRJ41NX6Z0: Downloading 1 format(s): hls-6101+hls-audio-stereo-aac-English
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 69
[download] Destination: Die Weltmeister-Werkstatt 1 [AA7U8970E8DRJ41NX6Z0].fhls-6101.mp4
[download]   4.6% of ~122.63MiB at    3.04MiB/s ETA 00:31 (frag 3/69)^C
ERROR: Interrupted by user
```

StefanLobbenmeier commented:
I might have found a non working example: https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/, but it could also be my build

```
~/Downloads ❯❯❯ ./yt-dlp_macos_legacy-2 -v https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/
[debug] Command-line config: ['-v', 'https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.09.140503 [9e642a8] (darwin_exe)
[debug] Python 3.10.5 (CPython 64bit) - macOS-12.3-x86_64-i386-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.0.1 (setts), ffprobe 5.0.1, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [Servus] Extracting URL: https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/
[Servus] AAVUYK29A8FSW6CSX4V7: Downloading video JSON
[Servus] AAVUYK29A8FSW6CSX4V7: Downloading m3u8 information
ERROR: 'bytes' object has no attribute 'encode'
Traceback (most recent call last):
  File "yt_dlp/YoutubeDL.py", line 1425, in wrapper
  File "yt_dlp/YoutubeDL.py", line 1495, in __extract_info
  File "yt_dlp/extractor/common.py", line 640, in extract
  File "yt_dlp/extractor/servus.py", line 55, in _real_extract
  File "yt_dlp/extractor/common.py", line 2132, in _extract_m3u8_formats_and_subtitles
  File "yt_dlp/extractor/common.py", line 834, in _download_webpage_handle
  File "yt_dlp/extractor/common.py", line 773, in _request_webpage
  File "yt_dlp/extractor/common.py", line 739, in _create_request
  File "yt_dlp/utils.py", line 752, in sanitized_Request
  File "yt_dlp/utils.py", line 2989, in escape_url
AttributeError: 'bytes' object has no attribute 'encode'
```

StefanLobbenmeier commented:
The example https://www.servustv.com/unterhaltung/v/aavuyk29a8fsw6csx4v7/ actually doesn’t play the full movie in the browser either, just a trailer saying that it will run at 20:15

So it might actually work if we try again later, but it would be nicer to have a dedicated error message or download the trailer? Not sure, but not a huge issue if it ends up working later I think

pukkandan commented:
How is it that there are so many people willing to make PR for this site, but no one want to address reviews... https://github.com/yt-dlp/yt-dlp/pull/1045 https://github.com/yt-dlp/yt-dlp/pull/2748 https://github.com/yt-dlp/yt-dlp/pull/1498

FrankZ85 commented:
> How is it that there are so many people willing to make PR for this site, but no one want to address reviews... #1045 #2748 #1498

I have just added some more functionality according to my research on the site, and addressed the reviews.

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+            self._report_errors(video)
+            return None
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
+            video['videoUrl'], video_id, 'mp4', entry_protocol='m3u8_native',
+            m3u8_id='hls')
```
```suggestion
            video['videoUrl'], video_id, 'mp4', m3u8_id='hls')
```
It is the default

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+
+    def _report_errors(self, video):
+        if 'playabilityErrors' not in video:
+            self.report_warning('No videoUrl, and also no information about errors')
```
raise ExtractorError. Same below. When the error does not indicate a bug (eg: `only playable from ...`), set `expected=true`

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
         }
+
+    def _get_description(self, video, video_id):
+        info = self._download_json("https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/%s?fieldset=page" % video_id,
+                                   video_id, fatal=False)
```
I don't think a description is useful enough to make an additional API call for it. But I leave it to your discretion

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+            'subtitles': subtitles,
         }
+
+    def _get_description(self, video, video_id):
+        info = self._download_json("https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/%s?fieldset=page" % video_id,
+                                   video_id, fatal=False)
+        if not info:
+            return video.get('description')
+        description = info.get('stv_long_description') \
+            or info.get("stv_short_description") \
+            or video.get('description')
+        return description
```
```suggestion
        return traverse_obj(
            info, 'stv_long_description', 'stv_short_description') or video.get('description')
```

pukkandan replied:
Actually, passing `video` here only to do a `.get` on it seems unnecessary. Do this instead:
```suggestion
    def _get_description(self, video_id):
        info = self._download_json(
            f'https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/{video_id}?fieldset=page',
            video_id, fatal=False)
        return traverse_obj(info, 'stv_long_description', 'stv_short_description')
```
and above
```py
            'description': self._get_description(video_id) or video.get('description'),
```

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+            'thumbnail': video.get('poster'),
+            'duration': float_or_none(video.get('duration')),
+            'timestamp': unified_timestamp(video.get('currentSunrise')),
+            'series': season,
```
typo?

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+            'duration': float_or_none(video.get('duration')),
+            'timestamp': unified_timestamp(video.get('currentSunrise')),
+            'series': season,
+            'season': video.get('season'),
```
```suggestion
            'season': season,
```

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
-                    'width': int_or_none(resource.get('width')),
-                    'height': int_or_none(resource.get('height')),
-                })
+            'https://api-player.redbull.com/stv/servus-tv?timeZone=Europe/Berlin&videoId=%s' % video_id,
```
fstrings are prefered over `%`, but I won't force you

pukkandan replied:
Use `query` param

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
         'only_matching': True,
     }, {
-        'url': 'https://www.pm-wissen.com/videos/aa-24mus4g2w2112/',
+        'url': 'https://www.servustv.com/natur/v/aansszcx3yi9jmlmhdc1/',
```
The existing tests are meant to test those URL schemes. Don't replace them

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
         }
+
+    def _get_description(self, video, video_id):
+        info = self._download_json("https://backend.servustv.com/wp-json/rbmh/v2/media_asset/aa_id/%s?fieldset=page" % video_id,
```
Use single quotes everywhere

pukkandan commented on yt_dlp/extractor/servus.py:
```diff
@@ -15,131 +12,98 @@ class ServusIE(InfoExtractor):
+            if error == 'FSK_BLOCKED':
+                details = video['playabilityErrorDetails']['FSK_BLOCKED']
+                if 'minEveningHour' in details:
+                    self.report_warning('Only playable from '
+                                        + f'{details["minEveningHour"]}:00 to '
+                                        + f'{details["maxMorningHour"]}:00')
```
format it like:
```py
                    raise ExtractorError(
                        f'Only playable from {details["minEveningHour"]}:00 to {details["maxMorningHour"]}:00',
                        expected=True)
```

gamer191 commented:
@flashdagger after this gets merged are you planning on maintaining your servustv plugin?

pukkandan commented:
@gamer191 It implements much more functionality afair

StefanLobbenmeier commented:
Is there anything I can do to make this PR merged? @FrankZ85 would you mind if I took it over

pukkandan commented:
> Is there anything I can do to make this PR merged?

If OP is not going to address the review, you can make a new PR addressing the current comments.

When making PR, make sure you are willing and able to address future reviews. I already reviewed 4 different PRs for the same issue, all the contributors abandoning the PR midway! If you don't want to deal with our conventions/requirements, distribute it as a plugin instead and pls don't waste time of all involved

StefanLobbenmeier commented:
> If OP is not going to address the review, you can make a new PR addressing the current comments.

We waited another 2 weeks after pinging him, so I am assuming he will not continue.

I will start by making a copy of this PR and replicate your PR comments, and then fix them in my version to minimize the work on your side 😄 as for dealing with new review comments I will deal with them
